src/package-constraint-resolver.js

```diff
@@ -5,20 +5,211 @@ import type Config from './config.js';
 // the complexity and is very efficient for package resolution.
 
 export default class PackageConstraintResolver {
-  constructor(config: Config, reporter: Reporter) {
+  constructor(config: Config, reporter: Reporter, resolver: PackageResolver) {
     this.reporter = reporter;
     this.config = config;
+    this.resolver = resolver
+
+    // cache of all package metadata, mapped from name to manifest
+    this.packageMetadata = {}
+    this.currentlyFetching = {}
+
+    // logic solver
+    this.logicSolver = new Logic.Solver()
   }
 
   reporter: Reporter;
   config: Config;
+  resolver: PackageResolver;
+
+  packageMetadata: Object;
+  currentlyFetching: Object;
 
   reduce(versions: Array<string>, range: string): Promise<?string> {
     return Promise.resolve(semver.maxSatisfying(versions, range, this.config.looseSemver));
   }
+
+  // ALl this method is used for currently is for fetching all package metadata
+  // It is called in the pipeline in place of Package Request's findVersionInfo
+  // so that we can retain the package metadata
+  async addPackage(req: DependencyRequestPattern, isTld?: boolean = false) {
+    const request = new PackageRequest(req, this.resolver);
+
+    // find the version info
+
+    const info: ?Manifest = await request.getPackageMetadata();
+
+    if (!info) {
+      throw new MessageError(this.reporter.lang('unknownPackage', req.pattern));
+    }
+
+    this.packageMetadata[info.name] = info
+
+    const dependenciesToFetch = []
+    // create new find's for all the dependencies
+    for (const ver in info.versions) {
+      const dependencyNames = Object.keys(info.versions[ver].dependencies || {})
+      dependencyNames.map((name) => {
+        // if we have metadata already, ignore. otherwise, call find again
+        if(this.packageMetadata[name] == null && this.currentlyFetching[name] == null) {
+          dependenciesToFetch.push(this.resolver.find({
+            // TODO: need to make it so it can just take a name instead of a
+            // pattern, because we're not looking for a specific version here
+            pattern: name,
+            // TODO
+            // registry: remote.registry,
+            registry: 'npm',
+            visibility: USED_VISIBILITY,
+            optional: true,
+            parentRequest: request,
+          }))
+          this.currentlyFetching[name] = true
+        }
+      })
+    }
+    // fetch all dependencies metadata for all versions
+    return await Promise.all(dependenciesToFetch);
+
+  }
+
+
+  // given a package name and a version, create a logic term
+  _createLogicTerm(name, ver) {
+    return `${name} ${ver}`
+  }
+
+  // given a logic term, generate the package name and version it represents
+  // in the form [name, version]
+  _parseLogicTerm(term) {
+    return term.split(' ')
+  }
+
+
+  // first call of all top level dependencies. We have to differentiate these
+  // packages because theY MUST be required
+  solve(topLevelDependencies): void {
```

Comment:
this method (and `solvePackage` below) is the actual setup for the logic solver.